gms/gossiper.cc

```diff
@@ -2028,28 +2070,108 @@ future<> gossiper::wait_for_gossip(std::chrono::milliseconds initial_delay, std:
     });
 }
 
+bool operator==(const gossip_query_token_status_response& x, gossip_query_token_status_response& y) {
+    return x.nr_known_nodes == y.nr_known_nodes
+            && x.normal_token_hash == y.normal_token_hash
+            && x.bootstrap_token_hash == y.bootstrap_token_hash
+            && x.leaving_nodes_hash == y.leaving_nodes_hash
+            && x.status_hash == y.status_hash;
+}
+
 future<> gossiper::wait_for_gossip_to_settle() {
-    static constexpr std::chrono::milliseconds GOSSIP_SETTLE_MIN_WAIT_MS{5000};
-    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
+    logger.info("Waiting for gossip to settle with gossip query...");
     auto do_enable_features = [this] {
         return async([this] {
             if (!std::exchange(_gossip_settled, true)) {
                maybe_enable_features();
             }
         });
     };
-    if (force_after == 0) {
-        return do_enable_features();
-    }
-    return wait_for_gossip(GOSSIP_SETTLE_MIN_WAIT_MS, force_after).then([this, do_enable_features] {
```

Comment:
I do not want to respect it in the new method because it is a very dangerous option.

Reply:
It was a very dangerous option even before. We have it to be able to skip the sync. Why skip only the old method, and not the new method as well?

Reply:
Because old method is not safe any way. Th new method is designed to be save.

Reply:
We don't have this option because the old method is not safe. Skipping the old method makes things even less safe. We have this option to boot faster, in an unsafe way. So it should apply to the new method as well.

Reply:
> We don't have this option because the old method is not safe. Skipping the old method makes things even less safe.

I am not saying we have this option because the old method is not safe. I am saying if we the new method which is safe, we should not allow the legacy option to make it unsafe. 

>  We have this option to boot faster, in an unsafe way. So it should apply to the new method as well.

What's the point of to be unsafe just to boot faster?

Reply:
And if the point is to boot faster without unnecessary waits, with this patch it will boot faster and being safe. This makes the option unnecessary with the new method.

Reply:
The new method also waits. There is no guarantee that it will complete within a bounded time in all situations.

If the user sets `skip_wait_for_gossip_to_settle`, his intention is to skip the wait, and we should respect it regardless of how robust we think our algorithm is.

It could be used as a workaround for a bug in the new method discovered in the field.

Reply:
> The new method also waits. There is no guarantee that it will complete within a bounded time in all situations.

Why does complete in a bounded time is important?

> 
> If the user sets `skip_wait_for_gossip_to_settle`, his intention is to skip the wait, and we should respect it regardless of how robust we think our algorithm is.

Still, my question remains. What's the point of doing this which is unsafe? It's like to allow the node to boot with wrong token ring which can corrupt the data. 

> 
> It could be used as a workaround for a bug in the new method discovered in the field.

This is a much better justification.